granulate_utils/linux/ns.py

```diff
@@ -67,11 +68,110 @@ def resolve_proc_root_links(proc_root: str, ns_path: str) -> str:
+
+
+def _get_process_nspid_by_sched_files(process: Process) -> int:
+    # Old kernel (pre 4.1) doesn't have an NSpid field in their /proc/pid/status file
+    # Instead, we can look through all /proc/*/sched files from inside the process' pid namespace, and due to a bug
+    # (fixed in 4.14) the outer PID is exposed, so we can find the target process by comparing the outer PID
+
+    def _find_inner_pid() -> Optional[int]:
+        pattern = re.compile(r"\((\d+), #threads: ")  # Match example: "java (12329, #threads: 11)"
+
+        procfs = Path("/proc")
+        for procfs_child in procfs.iterdir():
+            is_process_dir = procfs_child.is_dir() and procfs_child.name.isdigit()
+            if not is_process_dir:
+                continue
+
+            try:
+                sched_file_path = procfs_child / "sched"
+                with sched_file_path.open("r") as sched_file:
+                    sched_header_line = sched_file.readline()  # The first line contains the outer PID
+            except (FileNotFoundError, ProcessLookupError):
+                # That's OK, processes might disappear before we get the chance to handle them
+                continue
+
+            match = pattern.search(sched_header_line)
+            if match is not None:
+                outer_pid = int(match.group(1))
+                if outer_pid == process.pid:
+                    return int(procfs_child.name)
+
+        return None
+
+    # We're searching `/proc`, so we only need to set our mount namespace
+    inner_pid = run_in_ns(["mnt"], _find_inner_pid, process.pid)
+    if inner_pid is not None:
+        if not process.is_running():  # Make sure the pid wasn't reused for another process
+            raise NoSuchProcess(process.pid)
+
+        return inner_pid
+
+    # If we weren't able to find the process' nspid, he must have been killed while searching (we only search
+    # `/proc/pid/sched` files, and they exist as long as the process is running (including zombie processes)
+    assert not process.is_running(), f"Process {process.pid} is running, but we failed to find his nspid"
+
+    raise NoSuchProcess(process.pid)
+
+
+def is_same_ns(process: Union[Process, int], nstype: str, process2: Union[Process, int] = None) -> bool:
+    if isinstance(process, int):
+        process = Process(process)
+    if isinstance(process2, int):
+        process2 = Process(process2)
+    elif process2 is None:
+        process2 = Process()  # `self`
+
+    is_the_same_ns = get_process_ns_inode(process, nstype) == get_process_ns_inode(process2, nstype)
+
+    # If one of the processes isn't running, we checked the wrong one
+    if not process.is_running():
+        raise NoSuchProcess(process.pid)
+    if not process2.is_running():
+        raise NoSuchProcess(process2.pid)
+
+    return is_the_same_ns
+
+
+def get_process_ns_inode(process: Process, nstype: str):
+    try:
+        return os.stat(f"/proc/{process.pid}/ns/{nstype}").st_ino
```

Comment:
One way this check can be improves: if we're checking for an `nstype` that doesn't exist in this kernel, this will raise, despite `is_same_ns`'s should actually treat it as "yes, they're the same namespace".

We should handle this case (e.g: `cgroup` namespace, not available in 3.10)